### PR TITLE
Clean up HLS unused state removal transformation

### PR DESCRIPTION
**Pull request by phate**
This PR cleans up the unused state removal transformation. It does the following:
1. Introduces unit tests such that I am not blind while cleaning up.
2. Fix use-after-free bugs in the transformation such that valgrind does not get a heart attack any longer.
3. Clean up the code such that I can easily refactor the RemoveInput, RemoveOutput, ... methods.
4. Document the shortcomings of the transformation in the unit tests.

Ultimately, this transformation could and should be replaced by InvariantValueRedirection and DeadNodeElimination, but that is not part of this PR.

### CODE REVIEW

**Review by haved — DISMISSED** [review]
The code looks a lot cleaner.

I agree that the transformations could be performed as a combination of value redirection and DNE, but this cleanup was very nice.

I'm a bit curious when you would want to remove unused lambda arguments.
As you point out, either the signature changes, or state edges are removed, which removes the nice property that the final 3 call node arguments line up with the final 3 function arguments. I'm however not familiar enough with `hls` to know why this might be useful.

**Comment by phate** [comment]
@sjalander and @davidmetz : As agreed yesterday, I will hold back on merging this until I get an updated version of the HLS backend from you.

**Review by phate — COMMENTED** [review]
@sjalander @davidmetz : Am I allowed to merge this piece of code now, or is there soemthing else coming that might conflict with this one?

**Comment by sjalander** [comment]
@phate You can merge it.

File: jlm/hls/backend/rvsdg2rhls/UnusedStateRemoval.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      newResults.push_back(substitutionMap.lookup(result->origin()));
+    }
+  }
+  auto newLambdaOutput = newLambda->finalize(newResults);
+
+  // TODO handle functions at other levels?
+  assert(lambdaNode.region() == lambdaNode.region()->graph()->root());
+  assert((*lambdaNode.output()->begin())->region() == lambdaNode.region()->graph()->root());
+
+  JLM_ASSERT(lambdaNode.output()->nusers() == 1);
```
[the remark this answers is not in the document]
Comment (phate) [reply jlm/hls/backend/rvsdg2rhls/UnusedStateRemoval.cpp:110]: Fixed to `JLM_ASSERT()`

File: tests/jlm/hls/backend/rvsdg2rhls/UnusedStateRemovalTests.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  rvsdg.add_export(result, { valueType, "f" });
+
+  // Act
+  jlm::hls::RemoveUnusedStates(*rvsdgModule);
+
+  // Assert
+  // This assert is only here so that we do not forget this test when we refactor the code
+  assert(thetaNode->ninputs() == 1);
+
+  // FIXME: This transformation is broken for theta nodes. For the setup above, it
```
[the remark this answers is not in the document]
Comment (phate) [reply tests/jlm/hls/backend/rvsdg2rhls/UnusedStateRemovalTests.cpp:118]: Yes, it is. The point of this PR is not to fix the broken transformations, but to (hopefully) keep the state as is while making it easier to restructure this code in the future.

**Comment by phate** [comment]
> I'm a bit curious when you would want to remove unused lambda arguments.
As you point out, either the signature changes, or state edges are removed, which removes the nice property that the final 3 call node arguments line up with the final 3 function arguments. I'm however not familiar enough with hls to know why this might be useful.

@haved : States have a physical manifestation in HLS. Thus, unused states in HLS are similar to dead code in that they do not contribute to the result, but take up resources. However, from the outside/callers, i.e., C program, perspective the states are uninteresting as they are only a structural means to sequentialize program parts. They do not have a physical manifestation. This means we can remove them from the kernal function (in order to save resources on the FPGA, etc.), while keeping the correctness of the program.

Unused states could also be removed from the RVSDG for normal programs, we just do not have a pass for it right now.

@sjalander can probably provide some more details.

**Comment by sjalander** [comment]
> States have a physical manifestation in HLS. Thus, unused states in HLS are similar to dead code in that they do not contribute to the result, but take up resources. However, from the outside/callers, i.e., C program, perspective the states are uninteresting as they are only a structural means to sequentialize program parts. They do not have a physical manifestation. This means we can remove them from the kernal function (in order to save resources on the FPGA, etc.), while keeping the correctness of the program.

The comment summarize the situation nicely. State edges are used for sending tokens between (potentially) dependent operations to sequentialize them. If the operations are anyway ensured to be performed in correct order, then the state edge is only a hardware overhead.

@davidmetz do you have anything to add?

**Review by haved — APPROVED** [review]
LGTM, your comments make sense, even though I'm still a little bit unsure about how the call nodes are affected by the changing signatures, I'm assuming this isn't a problem in HSL

**Comment by phate** [comment]
@haved : The call nodes are absolutely affected by the changing signatures and would need to be replaced by new call nodes with an adjusted signature. The transformation does not take care of this either. However, this is not a problem in this case as the transformation is only meant to take care of the kernel (lambda) of the HLS translation. This kernel lambda has before been extracted into its own RVSDG module and all function calls have been inlined (no way to support them in HLS). In other words, the transformation (as it is right now) is only meant for a single lambda node within a module that contains no calls whatsoever.

**Pull request merged**